### ISSUE TITLE
is_resource_modified and weak entity tags

### ISSUE THREAD

**Issue by mitsuhiko** [mitsuhiko/werkzeug#124]
There was a report that the behavior is not what would be expected in regards to weak etags. This should be checked.

**Comment by jab** [comment]
Hi @mitsuhiko, I believe I just hit this myself and had to work around it. Sharing my use case and results in case it helps.

Several of my responses — such as for `/`, `/login`, and `/signup` — should support conditional GETs *for non-logged-in users*. This is because for logged-in users, these requests should redirect to `/home`.

So the logic looks something like this:
```python
@app.route('/')
def index():
    if g.user:
        return redirect(url_for('home'))
    return conditional_resp('index.html')

def conditional_resp(tmpl):
    body = render_template(tmpl)
    resp = make_response(body)
    resp.add_etag()
    resp.cache_control.no_cache = True
    return resp.make_conditional(request)
```
(Note this follows the recommendations [here](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching#defining-optimal-cache-control-policy).)

However, I have CloudFlare in front of the app, and I noticed conditional GETs were not working as expected. I tracked it down to this issue. CloudFlare in many cases converts a strong ETag sent by the origin into a weak ETag. As a result, the etag that the browser will send back in the `If-None-Match` header will be weak. However, I observed that `resp.make_conditional(request)` considers the response modified when the If-None-Match etag is weak, since the etag added to the response by `add_etag` is strong, even though they are otherwise identical. Changing this to `resp.add_etag(weak=True)` actually causes CloudFlare to drop the Etag header completely, so that doesn't help. In any case, I asked CloudFlare about this, and they said according to http://tools.ietf.org/html/rfc7232#section-3.2, Werkzeug should be performing a weak comparison when comparing ETags for If-None-Match.

For now, I am working around this by stripping any `W/` from `request.environ['HTTP_IF_NONE_MATCH']` before doing the above, which makes the If-None-Match etag strong, and therefore `resp.make_conditional(request)` works as expected.

It looks like the change would have to go in `werkzeug.datastructures.Etags.contains_raw` to fix the root cause. If you can confirm, I would be happy to work up a PR, or help with any other testing needed.

**Comment by jab** [comment]
I realized that when you do `resp.add_etag(weak=True)`, Werkzeug incorrectly prefixes the etag with `w/` rather than `W/`, which causes strict reverse proxies such as CloudFlare to drop the ETag header entirely (since technically it's invalid). Fixed on my "jab-fix-weak-etags" branch: #862

Since that still doesn't fix `is_resource_modified`'s oversensitivity to etag weakness, I created a "jab-fix-if-none-match-comparison" branch off the "jab-fix-weak-etags" branch and fixed that there (#863).

If both #862 and #863 look good, you can just merge #863, and GitHub should automatically merge and close #862 as well.

**Comment by jab** [comment]
Update: Rebased both branches on top of latest master (see #863 and #866). These should now both be merged independently.
